packages/vite/src/node/plugins/chunkImportMap.ts

```diff
@@ -0,0 +1,117 @@
+import path from 'node:path'
+import type { OutputBundle, OutputChunk } from 'rollup'
+import MagicString from 'magic-string'
+import { getHash, normalizePath } from '../utils'
+import type { Plugin } from '../plugin'
+import type { ResolvedConfig } from '../config'
+import type { IndexHtmlTransformHook } from './html'
+
+const hashPlaceholderLeft = '!~{'
+const hashPlaceholderRight = '}~'
+const hashPlaceholderOverhead =
+  hashPlaceholderLeft.length + hashPlaceholderRight.length
+export const maxHashSize = 22
+// from https://github.com/rollup/rollup/blob/91352494fc722bcd5e8e922cd1497b34aec57a67/src/utils/hashPlaceholders.ts#L41-L46
+const hashPlaceholderRE = new RegExp(
+  // eslint-disable-next-line regexp/strict, regexp/prefer-w
+  `${hashPlaceholderLeft}[0-9a-zA-Z_$]{1,${
+    maxHashSize - hashPlaceholderOverhead
+  }}${hashPlaceholderRight}`,
+  'g',
+)
+
+const hashPlaceholderToFacadeModuleIdHashMap = new Map()
+
+function augmentFacadeModuleIdHash(name: string): string {
+  return name.replace(
+    hashPlaceholderRE,
+    (match) => hashPlaceholderToFacadeModuleIdHashMap.get(match) ?? match,
+  )
+}
+
+export function createChunkImportMap(
+  bundle: OutputBundle,
+  base: string = '',
+): Record<string, string> {
+  return Object.fromEntries(
+    Object.values(bundle)
+      .filter((chunk): chunk is OutputChunk => chunk.type === 'chunk')
+      .map((output) => {
+        return [
+          base + augmentFacadeModuleIdHash(output.preliminaryFileName),
+          base + output.fileName,
+        ]
+      }),
+  )
+}
+
+export function chunkImportMapPlugin(): Plugin {
+  return {
+    name: 'vite:chunk-importmap',
+
+    // If the hash part is simply removed, there is a risk of key collisions within the importmap.
+    // For example, both `foo/index-[hash].js` and `index-[hash].js` would become `assets/index-.js`.
+    // Therefore, a hash is generated from the facadeModuleId to avoid this issue.
+    renderChunk(code, _chunk, _options, meta) {
+      Object.values(meta.chunks).forEach((chunk) => {
+        const hashPlaceholder = chunk.fileName.match(hashPlaceholderRE)?.[0]
+        if (!hashPlaceholder) return
+        if (hashPlaceholderToFacadeModuleIdHashMap.get(hashPlaceholder)) return
+
+        hashPlaceholderToFacadeModuleIdHashMap.set(
+          hashPlaceholder,
+          getHash(chunk.facadeModuleId ?? chunk.fileName),
+        )
+      })
+
+      const codeProcessed = augmentFacadeModuleIdHash(code)
+      return {
+        code: codeProcessed,
+        map: new MagicString(codeProcessed).generateMap({
+          hires: 'boundary',
+        }),
```

Comment:
📝 For reviewers: This works as rollup calculates the chunk hash after the `renderChunk` hooks are run since Rollup 3.

Reference: https://www.youtube.com/watch?v=cFwO9UvDzfI